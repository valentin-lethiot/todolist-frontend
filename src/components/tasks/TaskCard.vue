<script setup lang="ts">
import { Task, TaskStatus } from '@/models/tasks.entity'
import TaskCheckbox from '../forms/TaskCheckbox.vue'
import StateSelect from '../forms/StateSelect.vue'
import EditIcon from '../ui/EditIcon.vue'
import { useTasksStore } from '@/stores/tasks.store'

const { updateTask } = useTasksStore()

type Props = {
  task: Task
}
const { task } = defineProps<Props>()

const emit = defineEmits(['editTask'])

const emitEditTask = () => emit('editTask')

const updateTaskStatus = (status: TaskStatus) => {
  updateTask(task.id!, { ...task, status: status as TaskStatus })
}
</script>

<template>
  <div class="rounded-xl w-100 p-4 bg-gray-800 border-gray-700">
    <div class="flex flex-col">
      <div class="flex justify-between">
        <p class="w-5/6 text-lg font-bold text-white break-all">
          {{ task.title }}
        </p>
        <div>
          <button
            class="rounded-full p-2 hover:bg-gray-600 text-lg font-bold text-white hover:cursor-pointer"
            @click="emitEditTask"
          >
            <EditIcon />
          </button>
        </div>
      </div>
      <p class="text-gray-300">
        {{ task.description }}
      </p>
      <div class="flex mt-2 justify-end">
        <StateSelect label="Status" v-model="task.status" @update:model-value="updateTaskStatus" />
      </div>
    </div>
  </div>
</template>
