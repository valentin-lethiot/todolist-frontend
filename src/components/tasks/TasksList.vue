<script setup lang="ts">
import { Task } from '@/models/tasks.entity'
import TaskCard from '@/components/tasks/TaskCard.vue'

type Props = {
  tasks: Task[]
}
defineProps<Props>()
</script>

<template>
  <div v-for="task in tasks" :key="task.id" class="flex flex-col my-1">
    <TaskCard :task="task" />
  </div>
</template>
