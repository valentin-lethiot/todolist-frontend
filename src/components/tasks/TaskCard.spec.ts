import { describe, beforeEach, expect, test } from 'vitest'
import { mount, VueWrapper } from '@vue/test-utils'
import TaskCard from './TaskCard.vue'
import { Task, TaskStatus } from '@/models/tasks.entity'
import { createPinia, setActivePinia } from 'pinia'

describe('TaskCard', () => {
  let wrapper: VueWrapper

  const mockTask = new Task('Test Task', 'Test Description', TaskStatus.TODO, '1')

  beforeEach(() => {
    setActivePinia(createPinia())
    wrapper = mount(TaskCard, {
      props: {
        task: mockTask,
      },
    })
  })

  describe('render', () => {
    describe('desktop', () => {
      test('should render', () => {
        expect(wrapper.html()).toMatchSnapshot()
      })
    })
  })

  describe('actions', () => {
    test('should emit editTask event when edit button is clicked', async () => {
      await wrapper.find('#editButton').trigger('click')
      expect(wrapper.emitted('editTask')).toBeTruthy()
    })

    test('should update task status when state is changed', async () => {
      const stateSelect = wrapper.findComponent({ name: 'StateSelect' })
      await stateSelect.vm.$emit('update:modelValue', TaskStatus.DOING)
    })
  })
})
