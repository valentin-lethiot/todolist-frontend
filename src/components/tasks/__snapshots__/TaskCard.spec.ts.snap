// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TaskCard > render > desktop > should render 1`] = `
"<div class="rounded-xl w-full p-4 bg-gray-800 border-gray-700">
  <div class="flex flex-col">
    <div class="flex justify-between">
      <p class="w-full md:w-5/6 text-md md:text-lg font-bold text-white break-all">Test Task</p>
      <div class="hidden md:w-1/6 md:flex md:justify-end md:items-start"><button id="editIcon" class="rounded-full p-2 hover:bg-gray-600 text-md md:text-lg font-bold text-white hover:cursor-pointer"><svg xmlns="http://www.w3.org/2000/svg" width="16" height="16" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round">
            <path d="M17 3a2.828 2.828 0 1 1 4 4L7.5 20.5 2 22l1.5-5.5L17 3z"></path>
          </svg></button></div>
    </div>
    <p class="text-gray-300">Test Description</p>
    <div class="flex mt-4 justify-between md:justify-end"><button id="editButton" class="md:hidden rounded p-1.5 bg-gray-900 hover:bg-gray-600 text-sm font-bold text-white hover:cursor-pointer">Modifier</button><select class="border text-sm rounded-lg p-1 border-gray-60 text-white hover:cursor-pointer bg-red-600" label="Status" value="TODO">
        <option selected="" class="bg-red-500" value="TODO">A faire</option>
        <option class="bg-orange-500" value="DOING">En cours</option>
        <option class="bg-green-500" value="DONE">TerminÃ©</option>
      </select></div>
  </div>
</div>"
`;
